lis2de12.c

```diff
@@ -0,0 +1,248 @@
+#define TIME_LATENCY		0x3C
+#define TIME_WINDOW			0x3D
+
+#define MODE_CONFIG_OFF		0x00
+#define MODE_CONFIG_ON		0x0F
+
+static __attribute((far)) int i2c_bus;
+static __attribute((far)) int i2c_address;
+static __attribute((far)) lis2de12_cb cb;
+static __attribute((far)) char data[126];//6*21 read 21 value from fifo
+static __attribute((far)) unsigned char tap;
+static __attribute((far)) unsigned char reg;
+static void write(unsigned char reg, unsigned char data);
+
+static void lis2de12_i2c_cb(int i2c_id, bool status) {
+	if (reg==CLICK_SRC)
+		cb((signed char)data[0], (signed char)data[2], (signed char)data[4], data[5]);
+	else{
+		reg = CLICK_SRC;
+		i2c_master_transfert_async(i2c_bus, i2c_address, &reg, 1, (unsigned char *) data+5, 1, lis2de12_i2c_cb);
+	}
+}
+
+static void lis2de12_i2c_cb_fifo(int i2c_id, bool status) {
+	
+	if (reg==(FIFO_READ_START|0x80)){//reg==CLICK_SRC
+		int sumx=0;
+		int sumy=0;
+		int sumz=0;
+		int i;
+		for(i=1;i<125;i+=6){
+			sumx+=data[i];
+			sumy+=data[i+2];
+			sumz+=data[i+4];
+		}
+		if (sumx<0)
+			sumx=-((-sumx)>>6); // correct negative value rounding
+		else
+			sumx=sumx>>6; // 21/64=0.328~=0.33 to not divide by 3 afterward ack specific for Thymio
+		
+		if (sumy<0)
+			sumy=-((-sumy)>>6); // correct negative value rounding
+		else
+			sumy=sumy>>6; // 21/64=0.328~=0.33 to not divide by 3 afterward ack specific for Thymio
+		
+		if (sumz<0)
+			sumz=-((-sumz)>>6); // correct negative value rounding
+		else
+			sumz=sumz>>6; // 21/64=0.328~=0.33 to not divide by 3 afterward ack specific for Thymio
+		
+		cb((signed char)sumx, (signed char)sumy, (signed char)sumz,tap);
+	
+	}else if(reg==CLICK_SRC){
+		reg = FIFO_READ_START|0x80; //most significant bit enable multiple read and automatic roll back in fifo mode
+		i2c_master_transfert_async(i2c_bus, i2c_address, &reg, 1, (unsigned char *) data, 126, lis2de12_i2c_cb_fifo);
+	}
+}
+
+void lis2de12_read_async(void) {
+	// Safety: If i2c is busy, ignore this
+	if (i2c_master_is_busy(i2c_bus))
+		return;
+
+	// Read XOUT, YOUT, ZOUT, status and fire the callback
+	reg = XOUT|0x80;
+	i2c_master_transfert_async(i2c_bus, i2c_address, &reg, 1, (unsigned char *) data, 5, lis2de12_i2c_cb);
+}
+
+void lis2de12_read_async_fifo(void) {
+	// Safety: If i2c is busy, ignore this
+	if (i2c_master_is_busy(i2c_bus))
+		return;
+	//read tap
+	reg = CLICK_SRC;
+	i2c_master_transfert_async(i2c_bus, i2c_address, &reg, 1, &tap, 1, lis2de12_i2c_cb_fifo);
+}
+
+static void write(unsigned char reg, unsigned char data) {
+	unsigned char array[2];
+	array[0] = reg;
+	array[1] = data;
+	i2c_master_transfert_block(i2c_bus, i2c_address,
+			array, 2, NULL, 0);
+}
+
+int lis2de12_init(int i2c, unsigned char address, lis2de12_cb ucb, int prio) {
+	char id;
+	i2c_bus = i2c;
+	i2c_address = address;
+	cb = ucb;
+
+	va_get(); // Enable Va LDO
+
+	clock_delay_us(1300);
+	reg=LIS2DE12_WHO_AM_I;
+	i2c_master_transfert_block(i2c_bus,i2c_address,&reg,1,&id,1);
+	if (id!=LIS2DE12_ID)
```

Comment:
Errr, you have a problem here .You are breaking the VA managmenent on all the old thymio. (ie, all the thymio with old accelerometer will start to consume more in sleep after this change). You might want to either do a va_put here OR change the mma7660 to not do the va_get (*and put a comment there to explain why*)
If you code the abstraction layer, you can put the VA management in this layer and then it will be clean (hint: you should really do this abstraction layer ...)

Reply:
ok thank I got the va_get() and va_put(). I will think about the abstraction layer.